### PR TITLE
PERF: do ndim validation before Block.__init__

### PR DESCRIPTION
**Pull request by jbrockmendel**

### CODE REVIEW

File: pandas/core/internals/blocks.py [hunk 1]
```diff
@@ -2262,9 +2213,63 @@ def get_block_type(values, dtype: Optional[Dtype] = None):
...
+            raise ValueError(
+                f"Wrong number of items passed {len(values)}, "
+                f"placement implies {len(placement)}"
+            )
+    elif ndim == 2 and len(placement) != 1:
+        # TODO(EA2D): special case unnecessary with 2D EAs
+        raise AssertionError("block.size != values.size")
+
+
+def extract_pandas_array(values, dtype, ndim):
```
Comment (jreback) [inline pandas/core/internals/blocks.py:2272]: ideally type & full docstring, shouldn't this live with extract_array?

File: pandas/core/internals/managers.py [hunk 2]
```diff
@@ -1610,6 +1610,8 @@ def get_slice(self, slobj: slice, axis: int = 0) -> SingleBlockManager:
 
         blk = self._block
         array = blk._slice(slobj)
+        if array.ndim > blk.values.ndim:
+            raise ValueError
```
Comment (jreback) [inline pandas/core/internals/managers.py:1614]: message?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/internals/managers.py (hunk 2 above)
Comment (jbrockmendel) [reply pandas/core/internals/managers.py:1614]: could add something in to be on the safe side, but this is only reached when called from Series.get_value, where it is caught and handled

File: pandas/core/internals/blocks.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/core/internals/blocks.py:2272]: > ideally type & full docstring

sure

> shouldn't this live with extract_array?

short answer: not sure yet.  i need to untangle the patching we do in test_numpy (xref #40021)

**Comment by jbrockmendel** [comment]
Updated with exception message, docstring, annotations

File: pandas/tests/internals/test_internals.py [hunk 3]
```diff
@@ -1318,17 +1318,19 @@ def test_make_block_no_pandas_array(block_maker):
...
-    assert result.dtype.kind in ["i", "u"]
-    assert result.is_extension is False
-
-    # ndarray, PandasDtype
-    result = block_maker(
-        arr.to_numpy(), slice(len(arr)), dtype=arr.dtype, ndim=arr.ndim
-    )
-    assert result.dtype.kind in ["i", "u"]
-    assert result.is_extension is False
+    if block_maker is make_block:
```
Comment (jreback) [inline pandas/tests/internals/test_internals.py:1321]: i guess could have the else path here as well to assert the public api?

**Review by jreback — CHANGES_REQUESTED** [review]

**Review by jreback — APPROVED** [review]

File: pandas/tests/internals/test_internals.py (hunk 3 above)
Comment (jbrockmendel) [reply pandas/tests/internals/test_internals.py:1321]: this is the pseudo-public api

Comment (jreback) [reply pandas/tests/internals/test_internals.py:1321]: right what i mean is what's the else path? (an internal path)?

Comment (jbrockmendel) [reply pandas/tests/internals/test_internals.py:1321]: correct.  the internal path no longer takes the dtype kwd checked below

**Comment by jbrockmendel** [comment]
anything else here?  several perf-improving follow-ups as we get validation out of the constructors

**Pull request merged**